lib/elixir/lib/string.ex

```diff
@@ -1898,16 +1898,27 @@ defmodule String do
 
   """
   @spec starts_with?(t, t | [t]) :: boolean
-  def starts_with?(string, []) when is_binary(string) do
-    false
+  def starts_with?(string, prefix) when is_binary(string) and is_binary(prefix) do
+    starts_with_string?(string, byte_size(string), prefix)
   end
 
   def starts_with?(string, prefix) when is_binary(string) and is_list(prefix) do
-    "" in prefix or Kernel.match?({0, _}, :binary.match(string, prefix))
+    string_size = byte_size(string)
+    Enum.any?(prefix, &starts_with_string?(string, string_size, &1))
```

Comment:
I decided to always use the recursion with `starts_with_string?` instead of `:binary.match`. I couldn't find an example where `:binary.match` would be faster. See the linked benchmark files for what I've tried.